Sources/Auth/Google/GoogleAuthProvider.swift

```diff
@@ -0,0 +1,102 @@
+
+public final class GoogleAuthProvider: NSObject {
+  private let config: Config
+  private weak var delegate: GoogleAuthProviderDelegate?
+  private weak var presentingViewController: UIViewController?
+  private let authProvider: GIDSignIn
+  
+  /// Class initializer with `config`, `presentingViewController` and optional `delegate`.
+  init(with config: Config, on presentingViewController: UIViewController, delegate: GoogleAuthProviderDelegate?) {
+    self.config = config
+    self.presentingViewController = presentingViewController
+    self.delegate = delegate
+    self.authProvider = GIDSignIn.sharedInstance
+    super.init()
+  }
+}
+
+// MARK: - Public Methods
+extension GoogleAuthProvider: GoogleAuthProvidable {
+  /// SignIn user.
+  ///
+  /// Will notify the delegate with the `Response` object on success or with `Error` on error.
+  public func signIn() {
+    guard !authProvider.hasPreviousSignIn() else {
+      authProvider.restorePreviousSignIn()
+      return
+    }
+    
+    guard let presentingVC = presentingViewController else {
+      delegate?.googleAuthProviderDidFail(with: .missingPresentingViewController)
+      return
+    }
+    
+    authProvider
+      .signIn(with: .init(clientID: config.clientId),
+              presenting: presentingVC) { [weak self] user, error in
+        switch (user, error) {
+        case (let signedInUser?, _):
+          signedInUser.authentication.do { [weak self] auth, error in
+            switch (auth, error) {
+            case (.some(let auth), _):
+              let userProfile = signedInUser.profile
+              let fullName = [userProfile?.givenName, userProfile?.familyName]
+                .compactMap { $0 }
+                .joined(separator: " ")
+              let response = AuthProvider.Response(token: auth.accessToken,
+                                                   name: fullName,
+                                                   email: userProfile?.email)
+              self?.delegate?.googleAuthProviderDidSignIn(with: response)
+            case (_, .some(let error)):
+              self?.delegate?.googleAuthProviderDidFail(with: .system(error))
+            default:
+              return
+            }
+          }
+        case (_, let actualError?):
+          let errorCode = (actualError as NSError).code
+          guard errorCode != GIDSignInError.Code.canceled.rawValue else { return }
+          self?.delegate?.googleAuthProviderDidFail(with: .system(actualError))
+        case (.none, .none):
+          self?.delegate?.googleAuthProviderDidFail(with: .unhandledAuthorization)
+        }
+      }
+  }
+  
+  /// Clears the signIn footprint and logs out the user immediatelly.
+  public static func signOut() {
+    GIDSignIn.sharedInstance.signOut()
+  }
+  
+  /// Checks the current auth state and returns the boolean value.
+  public static func checkAuthState(_ state: @escaping (Bool) -> Swift.Void) {
+    state(GIDSignIn.sharedInstance.currentUser != nil)
```

Comment:
Is this actually `@escaping`?

Reply:
It is, but only for Apple provider since method `getCredentialState` is escaping.

Reply:
Do we have plans for adopting more of swift concurrency instead of closures?

Reply:
Not atm. You are welcome to extend it 💪

Reply:
We wan still return a Promise instead of using a closure. 💁🏼